blog/2022/04/simplechains.md

```diff
@@ -0,0 +1,493 @@
+
+## SimpleChains.jl in Action: 10x-ing PyTorch in Small Examples
+
+Let's test MNIST with LeNet5. Note that this example will be a very conservative estimate of speed because, as a more traditional machine learning use case, batching can be used to make use of matrix-matrix multiplications instead of the even smaller matrix-vector kernels. That said, even in this case we'll be able to see a substantial performance benefit because of the semi-small network sizes.
+
+The following is the Julia code using SimpleChains.jl for the training:
+
+```julia
+using SimpleChains, MLDatasets
+
+lenet = SimpleChain(
+  (static(28), static(28), static(1)),
+  SimpleChains.Conv(SimpleChains.relu, (5, 5), 6),
+  SimpleChains.MaxPool(2, 2),
+  SimpleChains.Conv(SimpleChains.relu, (5, 5), 16),
+  SimpleChains.MaxPool(2, 2),
+  Flatten(3),
+  TurboDense(SimpleChains.relu, 120),
+  TurboDense(SimpleChains.relu, 84),
+  TurboDense(identity, 10),
+)
+
+# 3d and 0-indexed
+xtrain3, ytrain0 = MLDatasets.MNIST.traindata(Float32);
+xtest3, ytest0 = MLDatasets.MNIST.testdata(Float32);
+
+xtrain4 = reshape(xtrain3, 28, 28, 1, :);
+xtest4 = reshape(xtest3, 28, 28, 1, :);
+
+ytrain1 = UInt32.(ytrain0 .+ 1);
+ytest1 = UInt32.(ytest0 .+ 1);
+
+lenetloss = SimpleChains.add_loss(lenet, LogitCrossEntropyLoss(ytrain1));
+
+# initialize parameters
+@time p = SimpleChains.init_params(lenet);
+
+# initialize a gradient buffer matrix; number of columns places an upper bound
+# on the number of threads used.
+G = similar(p, length(p), min(Threads.nthreads(), (Sys.CPU_THREADS ÷ ((Sys.ARCH === :x86_64) + 1))));
+
+# train
+@time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+
+# assess training and test loss
+SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+
+
+# reinitialize to randomize parameters
+# SimpleChains.init_params!(lenet, p);
+
+# train without additional memor allocations
+@time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+
+# assess training and test loss
+SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+```
+
+#### PyTorch
+
+Before we show the results, let's look at the competition. Here's two runs of 10 epochs using PyTorch following [this script](https://github.com/chriselrod/LeNetTorch) on a V100 GPU using a batch size of 2048 produces:
+
+```
+Took: 17.33
+Accuracy: 0.9492
+Took: 17.22
+Accuracy: 0.9751
+```
+PyTorch using an AMD EPYC 7513 32-Core Processor:
+```
+Took: 14.86
+Accuracy: 0.9626
+Took: 15.09
+Accuracy: 0.9783
+```
+PyTorch using an Intel i9 10980XE 18-Core Processor:
+```
+Took: 11.24
+Accuracy: 0.9759
+Took: 10.78
+Accuracy: 0.9841
+```
+
+#### Flux.jl
+
+The standard machine learning library in Julia, [Flux.jl](https://github.com/FluxML/Flux.jl) was benchmarked using [this script]() with an A100 GPU. How fast was that?
```

Comment:
@chriselrod I don't have this script.

Reply:
https://github.com/PumasAI/SimpleChains.jl/blob/35d09c4ed3f72e43d73a2c391a6bbdceeb779e21/examples/mnist_lenet.jl

Reply:
https://github.com/PumasAI/SimpleChains.jl/blob/conv/examples/mnist_lenet.jl

Reply:
Eventually we'd need to `s/conv/main/g`, or aim it at a specific commit.
But for now the script is only on the conv branch, which is still in the unmerged PR.

Reply:
oof, still worth putting for now.